### PR TITLE
Remove dead code

### PR DESCRIPTION
**Pull request by EmilyBourne**
Pyccel still contains lots of dead code. This makes it harder to work on new developments as apparent dependencies are not always still relevant. This PR removes dead code and unused-imports. The unused elements are:

* Several elements in module `pyccel.ast.core`:
   - Method `ClassDef.get_attribute` (class attributes are collected from the scope)
   - Class `ForIterator` (mostly superseded by `Iterable` but hasn't been used since 2021 5a7cc8d542e336f3b0a1a390dfddf039172bafa2)
   - Function `get_iterable_ranges` (used in `ForIterator`)
   - Function `inline` (used for untested `With`. These functions don't have to be inline but can be marked with the decorator)
   - Function `subs` (used for `inline`)

* Some functions in `pyccel.ast.datatypes`:
   - `is_iterable_datatype`
   - `is_with_construct_datatype`
   - `is_pyccel_datatype`

### CODE REVIEW

File: pyccel/ast/core.py [hunk 1]
```diff
@@ -1008,8 +941,8 @@ def block(self):
                 start = i
             elif str(i.name) == '__exit__':
                 end   = i
-        start = inline(start,[])
-        end   = inline(end  ,[])
+        start = FunctionCall(start,[])
+        end   = FunctionCall(end  ,[])
 
         # TODO check if enter is empty or not first
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/core.py:948]: /bot accept
Known untested code. With sections need class magic methods

**Review by mkaddani — APPROVED** [review]
Looks good

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
+
+### Added
+
+### Fixed
+
+### Changed
+
+### Deprecated
+
+-   \[INTERNALS\] #1584 : Remove unused functions : `inline`, `subs`, `is_iterable_datatype`, `is_with_construct_datatype`, `is_pyccel_datatype`.
```
Comment (yguclu) [inline CHANGELOG.md:14]: I would specify the modules from which the functions have been removed. Further, I believe the function `get_iterable_ranges` is missing.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+-   \[INTERNALS\] #1584 : Remove unused class `ForIterator`.
```
Comment (yguclu) [inline CHANGELOG.md:15]: Same comment: this was removed from `pyccel.ast.core`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,20 @@
...
+-   \[INTERNALS\] #1584 : Remove unused method `get_attribute`.
```
Comment (yguclu) [inline CHANGELOG.md:16]: Same comment as above. For clarity I would also mention the class: `ClassDef.get_attribute`.

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good! I have some minor suggestions.

File: CHANGELOG.md (hunk 4 above)
Comment (yguclu) [reply CHANGELOG.md]: 1c808dd

**Review by yguclu — APPROVED** [review]

**Pull request merged**